File: 101. Symmetric Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }
    
    return isSymmetry(root.Left, root.Right)
}

func isSymmetry(a *TreeNode, b *TreeNode) bool {
    if a == nil && b != nil {
        return false
    } else if a != nil && b == nil {
        return false
    } else if a == nil && b == nil {
        return true
    } else if a.Val != b.Val {
        return false
    } else {
        return (isSymmetry(a.Left, b.Right) && isSymmetry(b.Left, a.Right))
    }
}

File: 101. Symmetric Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        
        return self.myIsSymmetric(root.left, root.right)
        
    def myIsSymmetric(self, left, right):
        if not left and not right:
            return True
        if not right and left:
            return False
        if right and not left:
            return False
        
        if left.val == right.val:
            return  self.myIsSymmetric(left.left, right.right) and \
                    self.myIsSymmetric(right.left, left.right)
        else:
            return False
